# 1. Exercise wrangling Procedures

In [73]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [74]:
df_ords = pd.read_csv(r'/Users/renataherrera/Documents/CF RH 2023-2024/CF DATA IMMERSION/CF RH A4 PYTHON/RH_PYTHON_Instacart Basket Analysis/02 Data/Original Data/orders.csv', index_col = False)

In [75]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

Dropping column - using pandas to manipulate data within a df that you've already imported into Python

In [76]:
# Dropping eval_set column from "orders.csv" (prior data set released by IC)
# prints head of the df without "eval_set" column
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

Using the value_counts() function on the "days_since_prior_order" columns in your df_ords dataframe and spot the missing (NaN) value

In [77]:
# Using "value_counts" to print all the functions within a specific column 
# AND to display multiple missing values as NaN
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

Renaming obscure or unituitive columns in df (Note: always check with client, re: may be naming conventions based on original dataset)

In [78]:
# Rename order_dow 
# and overwrite the original column name with this new column; rather than creating a new copy
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [79]:
# Check if code executed successfully
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

Changing a Variable's Data Type to perform a function on relevant columns and to something the function (eg. describe), will ignore, ie. a string

In [80]:
# order_id data type changed to a string
# and overwrite the existing "order_id" column with a new string version of the same column
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [81]:
# function to return the data type of your new "order_id" column
# an "O", which stands for object, Pandas' version of the string
df_ords['order_id'].dtype

dtype('O')

# Importing and transposing a new dataset to create a new df for it

In [82]:
# assign a path 
path = r'/Users/renataherrera/Documents/CF RH 2023-2024/CF DATA IMMERSION/CF RH A4 PYTHON/RH_PYTHON_Instacart Basket Analysis'

In [83]:
path

'/Users/renataherrera/Documents/CF RH 2023-2024/CF DATA IMMERSION/CF RH A4 PYTHON/RH_PYTHON_Instacart Basket Analysis'

In [84]:
# Import departments.csv data set
# and create a new df for it
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [85]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [86]:
# T function to transpose your data from "wide format" into "long format"
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [87]:
# overwrite the old df_dep df with the new transposed df_dep df
# create a new df from the transposed version and call it df_dep_t
df_dep_t = df_dep.T

In [88]:
# check
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [89]:
# add an index to a df that doesn't have one
# the index column lets you quickly identify rows by number
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

# Creating a new header and using the "iloc[ ]" function 

Used for selecting data in pandas and takes the index number of a row or column (or multiple rows and columns).By executing and assigning it to a variable, the values within that row(s) or column(s)are assigned to that variable.

In [90]:
# tell Python to take the row under the current header (the first row in the df)
# use that data to create a new header, and get rid of the current header
new_header = df_dep_t.iloc[0]

In [91]:
# check
new_header

0    department
Name: department_id, dtype: object

In [92]:
# create a new dataframe that only copies over rows beyond the first row
# Take the data under the header row for a new df
df_dep_t_new = df_dep_t[1:]

What does that 1: mean? This is telling Python to copy everything from the first row onward from the df_dep_t dataframe into a new dataframe, df_dep_t_new. The 1 represents row 1 in the dataframe. Remember that since Python starts indexing from 0 rather than 1, “row 1” is actually the second row in the dataframe. The colon that follows the 1 represents onward until the end of the dataframe. Together, they tell Python to copy everything starting from row 1 until the end of the dataframe.

In [93]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [94]:
# set your new_header variable as the new header for your df_dep_t_new dataframe
# tells Python to use the list of column names in your new_header variable as your new header
df_dep_t_new.columns = new_header

In [95]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

#Congrats to me, now my df is ready to be used as intended
#-as a DATA DICTIONARY!!

# Creating a data dictionary from "departments.csv" data set 

In [96]:
# use pandas and turn your df_dep_t_new dataframe into a dictionary
# and create a data dictionary that contains meanings for the values in the "department_id" column
# saved as a new variable, data_dict
data_dict = df_dep_t_new.to_dict('index')

In [97]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [98]:
#Import products.csv dataset
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [99]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [100]:
# Row 0 has a "department_id" of 19
# use your data dictionary to determine what this stands for
print(data_dict.get('19'))

{'department': 'snacks'}


# Subsetting - creating a smaller data set from a whole data set based on a particular filter

In [101]:
# create a subset for your df_prods df that only contains data from the snacks department
df_snacks =  df_prods[df_prods['department_id']==19]

In [102]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

Indexing refers to retrieving some element from an object, be it a dataframe, list, or otherwise.You're telling Python to look for the element in the brackets within the object to the left of the brackets (so, in this case, looking for a "department_id" value of 19 within the df_prods dataframe

In [103]:
# look for "department_id" value of 19 within the df_prods df
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [104]:
# save results in a new df_snacks df and check
df_snacks =  df_prods[df_prods['department_id']==19]

In [105]:
# check
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

#  A second way of achieving these same results--this involves use of the loc function - loc looks for a column name

iloc version - look for instances where the value of "department_id" is 19 within the df_prods database

In [106]:
# one version in using loc to extract subset
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

In [107]:
# check
df_snacks_2.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

Another iloc version - telling the loc function to look into a list: isin([19]). This is handy because you could use more than one value here if you wanted to check multiple "department_ids" at the same time (isin([17,18,19])).

In [108]:
# another version in using loc to extract subset
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

In [109]:
df_snacks_3.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# 2. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [110]:
# Check the data types of multiple columns in df_ords
df_ords.dtypes

order_id                   object
user_id                     int64
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [111]:
# Change user_id int64 data type to string
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [112]:
# Return the data type of new "user_id" column as "O" = string
df_ords['user_id'].dtype

dtype('O')

# 3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [113]:
#Renaming days_since_prior_order column without overwriting the df
df_ords.rename(columns = {'days_since_prior_order' : 'days_since_previous_order'}, inplace = True)

In [114]:
#Check
df_ords.head()

order_id user_id eval_set  order_number  orders_day_of_week  \
0  2539329       1    prior             1                   2   
1  2398795       1    prior             2                   3   
2   473747       1    prior             3                   3   
3  2254736       1    prior             4                   4   
4   431534       1    prior             5                   4   

   order_hour_of_day  days_since_previous_order  
0                  8                        NaN  
1                  7                       15.0  
2                 12                       21.0  
3                  7                       29.0  
4                 15                       28.0

# 3 A: days_since_previous_order changed from days_since_prior_order without overwrting the df

# 4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

df[`variable'] is a list of columns that you want to perform the value_counts() function on.
A series--a specific data structure in Pandas--a standard way of accessing a column within a df using Pandas

In [115]:
df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [116]:
# Only shows the column's existing values
# getting the frequency of a variable; "busiest hour for placing orders"
df['order_hour_of_day'].value_counts()

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

# 4 A: 10:00 am is the busiest hour with a frequency of 288418 placed orders

# 5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [117]:
# Using data dictionary to determine value of 4 in the "department_id" columnwithin the df_prods
print(data_dict.get('4'))

{'department': 'produce'}


In [118]:
#Check
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

# 5 A: Now you know that the meaning behind value 4 is  "produce" in that department. Fresh!

# 6. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.


Using loc looks for a column name. In this case, you want this function to look for instances where the value of "department_id" is 14 within the df_prods database.

In [119]:
#Using loc to create a subset containing only breakfast item sales info
df_breakfast = df_prods.loc[df_prods['department_id'] == 14]

In [120]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [121]:
# Another option for first 5 head rows subset of only breakfast item sales info
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

# 6 A: Depending on how big df is and how much "more" the sales team wants to know about breakfast item sales there are a couple of subset options to create.

# 7. They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [122]:
# Using loc and isin to create a subset and look into a list of more than one value
# and check multiple "department_ids" at the same time
df_dinnerpartyproducts = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [123]:
df_dinnerpartyproducts.head(50)

product_id                                       product_name  aisle_id  \
2             3               Robust Golden Unsweetened Oolong Tea        94   
6             7                     Pure Coconut Water With Orange        98   
9            10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10           11                                  Peach Mango Juice        31   
16           17                                  Rendered Duck Fat        35   
19           20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
22           23                             Organic Turkey Burgers        49   
34           35     Italian Herb Porcini Mushrooms Chicken Sausage       106   
38           39           Daily Tangerine Citrus Flavored Beverage        64   
39           40  Beef Hot Links Beef Smoked Sausage With Chile ...       106   
43           44                        Sparkling Raspberry Seltzer       115   
48           49      Vegetarian Grain Meat Sausages Italian - 4 CT        14   
51           52                                Mirabelle Brut Rose       134   
52           53              Whole Leaf Pure Aloe With Lemon Juice        98   
61           62            Premium Deli Oven Roasted Turkey Breast        96   
68           69                                                NaN        26   
72           73                  Jasmine Tea Unfiltered Ginger Ale        77   
73           74                    Artisan Chick'n & Apple Sausage        14   
81           82                       All Natural 100% Apple Juice        98   
83           84                                         Lamb Shank         7   
84           85                               Soppressata Piccante        96   
86           87                             Classics Earl Grey Tea        94   
96           97                        Organic Chamomile Lemon Tea        94   
108         109                             Grape Leaf Hummus Wrap        13   
118         119                             Chardonnay Paso Robles        62   
137         138                                 Olli Napoli Salami        96   
146         147                           Organic Ranch Veggie Dip        67   
149         150                                          Brut Rosé       134   
157         158  Smoked Ham & Oven Roasted Turkey Sub Sandwich Kit        96   
166         167                        Sparkling Blush Grape Juice        77   
174         175                                       T Bone Steak       122   
175         176                                    Original Turkey        96   
176         177                           Citrus Terere Yerba Mate        94   
187         188                                 Mild El Nino Salsa        67   
192         193        fruitwater® Strawberry Kiwi Sparkling Water        98   
193         194                                     Lamb Rib Chops       122   
195         196                                               Soda        77   
196         197                  Cold Brew Coffee Tahitian Vanilla        26   
197         198                            Tzatziki Cucumber Sauce        67   
208         209                                Italian Pasta Salad         1   
220         221      Cayenne Lemonade Made With Raw Tristate Honey        31   
222         223                    Kiwi Strawberry Sparkling Water       115   
229         230                                 Vanilla Cream Soda        77   
230         231                                  Low Sodium Turkey        49   
232         233                        Chinese Breakfast Black Tea        94   
233         234                                  Tennessee Whiskey       124   
236         237                            Brambleberry Herbal Tea        94   
248         249                     Pinot Grigio, California, 2010        62   
252         253      Organic Vegetable Greens & Greens Juice Blend        31   
256  

# 7 A: When executing the subset with less than 50 only 3 "alcohol" ( eg. "wine" types only and no other alcoholic specified) product details appear, so either I would suggest to run a separate products list to specify to include alcoholic/non-alcoholic or both for dinner parties depending on the client request. 

# 8. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [124]:
#Using shape to return dimensions of a df
df_dinnerpartyproducts.shape

(7650, 5)

# 8 A 7650 rows in dinnerpartyproducts df

# 9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

# 12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [125]:
#Import path
import os

In [126]:
#Create path
path = r'/Users/renataherrera/Documents/CF RH 2023-2024/CF DATA IMMERSION/CF RH A4 PYTHON/RH_PYTHON_Instacart Basket Analysis'

In [127]:
#Check path
path

'/Users/renataherrera/Documents/CF RH 2023-2024/CF DATA IMMERSION/CF RH A4 PYTHON/RH_PYTHON_Instacart Basket Analysis'

In [129]:
#Export wrangled df_ords into Prepared Data
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [131]:
# Create a subset to display extract of customer with a user_id of "1"
df_userid_1 = df_ords.loc[df_ords['user_id']=='1']

In [132]:
#Display extracted userid_1 customer info
df_userid_1

order_id user_id eval_set  order_number  orders_day_of_week  \
0   2539329       1    prior             1                   2   
1   2398795       1    prior             2                   3   
2    473747       1    prior             3                   3   
3   2254736       1    prior             4                   4   
4    431534       1    prior             5                   4   
5   3367565       1    prior             6                   2   
6    550135       1    prior             7                   1   
7   3108588       1    prior             8                   1   
8   2295261       1    prior             9                   1   
9   2550362       1    prior            10                   4   
10  1187899       1    train            11                   4   

    order_hour_of_day  days_since_previous_order  
0                   8                        NaN  
1                   7                       15.0  
2                  12                       21.0  
3                   7                       29.0  
4                  15                       28.0  
5                   7                       19.0  
6                   9                       20.0  
7                  14                       14.0  
8                  16                        0.0  
9                   8                       30.0  
10                  8                       14.0

# 10. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [133]:
# Statistical summary for "userid_1"
df_userid_1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_previous_order  
count                  10.000000  
mean                   19.000000  
std                     9.030811  
min                     0.000000  
25%                    14.250000  
50%                    19.500000  
75%                    26.250000  
max                    30.000000

# 13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [162]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'))